Frameworks/CoreGraphics/CGPath.mm

```diff
@@ -565,32 +584,66 @@ void CGPathAddArc(CGMutablePathRef path,
                   bool clockwise) {
     RETURN_IF(!path);
 
-    CGPoint startPoint = CGPointMake(x + radius * cos(startAngle), y + radius * sin(startAngle));
-    CGPoint endPoint = CGPointMake(x + radius * cos(endAngle), y + radius * sin(endAngle));
+    // Normalize the angles to work with to values between 0 and 2*PI
+    CGFloat normalizedStartAngle = __normalizeAngle(startAngle);
+    CGFloat normalizedEndAngle = __normalizeAngle(endAngle);
+
+    CGPoint startPoint = CGPointMake(x + radius * cos(normalizedStartAngle), y + radius * sin(normalizedStartAngle));
+    CGPoint endPoint = CGPointMake(x + radius * cos(normalizedEndAngle), y + radius * sin(normalizedEndAngle));
 
     // Create the parameters for the AddArc method.
     const D2D1_POINT_2F endPointD2D = _CGPointToD2D_F(endPoint);
     const D2D1_SIZE_F radiusD2D = { radius, radius };
-    CGFloat rotationAngle = abs(startAngle - endAngle);
     D2D1_ARC_SIZE arcSize = D2D1_ARC_SIZE_SMALL;
-    CGFloat expectedAngle = (clockwise ? startAngle + rotationAngle : startAngle - rotationAngle);
 
     // D2D does not understand that the ending angle must be pointing in the proper direction, thus we must translate
     // what it means to have an ending angle to the proper small arc or large arc that D2D will use since a circle will
     // intersect that point regardless of which direction it is drawn in.
-    if (expectedAngle == endAngle) {
-        arcSize = D2D1_ARC_SIZE_LARGE;
+    CGFloat rawDifference = 0;
+    if (clockwise) {
+        rawDifference = normalizedStartAngle - normalizedEndAngle;
     } else {
-        rotationAngle = (2 * M_PI) - rotationAngle;
+        rawDifference = normalizedEndAngle - normalizedStartAngle;
+    }
+    CGFloat difference = rawDifference;
+    if (difference < 0) {
+        difference += 2 * M_PI;
     }
+    if (difference > M_PI) {
+        arcSize = D2D1_ARC_SIZE_LARGE;
+    }
+
+    rawDifference = abs(rawDifference);
+    if (!clockwise) {
+        rawDifference *= -1;
+    }
+
+    // The direction of the arc's sweep must be reversed since the coordinate systems for D2D and CoreGraphics are reversed.
+    // CW in D2D is CCW in CoreGraphics.
     D2D1_SWEEP_DIRECTION sweepDirection = { clockwise ? D2D1_SWEEP_DIRECTION_COUNTER_CLOCKWISE : D2D1_SWEEP_DIRECTION_CLOCKWISE };
-    D2D1_ARC_SEGMENT arcSegment = D2D1::ArcSegment(endPointD2D, radiusD2D, rotationAngle, sweepDirection, arcSize);
 
     ComPtr<ID2D1PathGeometry> newPath;
     ComPtr<ID2D1GeometrySink> newSink;
+    if (!path->IsFigureOpen()) {
+        CGPathMoveToPoint(path, transform, startPoint.x, startPoint.y);
+    }
     FAIL_FAST_IF_FAILED(_createPathReadyForFigure(path, startPoint, &newPath, &newSink));
 
-    newSink->AddArc(arcSegment);
+    // This will only happen when drawing a circle in the clockwise direction from 2pi to 0, a scenario
+    // supported on the reference platform.
+    if (abs(abs(rawDifference) - 2 * M_PI) < .00001) {
```

Comment:
But could still be negative from a counterclockwise rotation. This negative is important later on but the strict value being close enough to zero here is the important factor for when to draw a full circle.